lib/machines.js

```diff
@@ -1113,6 +1114,93 @@ function dnsify(str) {
     return (str.toLowerCase().replace(/[^a-z0-9-]+/g, '-'));
 }
 
+// blatently stolen from sdc-vmapi lib/common/validation.js
+
+function validVolumeName(volumeName) {
+    var VALID_VOLUME_NAME_REGEXP = /^[a-zA-Z0-9][a-zA-Z0-9_\.\-]+$/;
+    if (volumeName === undefined || volumeName === null ||
+        !VALID_VOLUME_NAME_REGEXP.test(volumeName)) {
+        return false;
+    }
+
+    return true;
+}
+
+function validVolumeMountpoint(mountpoint) {
+    if (typeof (mountpoint) !== 'string') {
+        return false;
+    }
+
+    /*
+     * Must start with '/'.
+     */
+    if (mountpoint.length === 0 || mountpoint[0] !== '/') {
+        return false;
+    }
+
+    /*
+     * Must not contain the null character.
+     */
+    if (mountpoint.indexOf('\0') !== -1) {
+        return false;
+    }
+
+    /*
+     * Must contain at least one character that is not '/'.
+     */
+    if (mountpoint.search(/[^\/]/) === -1) {
+        return false;
+    }
+
+    return true;
+}
+
+// stops at first error
+function validateVolume(volume) {
+    var VALID_VOLUME_MODES = ['ro', 'rw'];
+
+    if (volume.type !== undefined && volume.type !== 'tritonnfs') {
+        return (new InvalidArgumentError('unknown volume type "' + volume.type
```

Comment:
If we return on the first error, then we lose the opportunity to report all relevant errors for a given request. Should we instead return an array of errors?

Reply:
Sure, I'll change to do that.